Add "Mutual-TLS"/"Client-Cert"-Authentication example to wildfly-camel-examples

jochenr opened the pull request:
Hello James,

follow up to [Pull Request 47](https://github.com/wildfly-extras/wildfly-camel-examples/pull/47), I added an example that does mutual-TLS / Client-Cert Authentication.
Possible I should change the name to meet the use-case better, but I fist wanted you to reviwe it, before I put more work into it.

Best regards,
Jochen

jamesnetherton commented on itests/pom.xml:
```diff
@@ -53,7 +53,13 @@
+
+        <dependency>
+			<groupId>org.wildfly</groupId>
+			<artifactId>wildfly-testsuite-shared</artifactId>
```
I need to check this. We used to have this in the test suite of the main project, but I removed it because it brought in  a bucket load of other dependencies.

If there's stuff we need from this dependency, maybe we recreate it here (if it's not a huge piece of work to replicate).

jochenr replied:
I thought that this doiesn't matter a lot, since I use the dependency in "provided" scope (and I didn't even have to use a version, since that was alreday there by inheritance).

As you see in the integration I only need three classes:
```
@Deployment
    public static WebArchive createDeployment() {
        return ShrinkWrap.createFromZipFile(WebArchive.class, new File("target/examples/example-camel-cxf-jaxws-cdi-secure.war"))
                .addClass(ContainerResourceMgmtTestBase.class)
                .addClass(org.jboss.as.test.integration.management.base.AbstractMgmtTestBase.class)
                .addClass(ModelUtil.class);
    }
```

And if I switch to the other DMRUtil perhaps only two...

jamesnetherton replied:
Understood, but scope is not the issue. It's the additional transitive dependencies that I don't like. I want to keep this project lightweight. Just as a comparison of additional WildFly dependencies that this thing brings in:

**Without testsuite-shared:**

```
$ mvn dependency:tree  | grep wildfly | grep -v camel | wc -l
51
```

**With testsuite-shared:**

```
$ mvn dependency:tree  | grep wildfly | grep -v camel | wc -l
156
```

I'll ponder this some more. For EAP / JBoss Fuse we don't actually ship the 'itests' module in the final distribution. So maybe this isn't such an issue. But I'd still favour keeping things as lightweight as is possible.

jochenr replied:
it was easier than I thought.
I removed the dependency on "org.wildfly:wildfly-testsuite-shared" and just addded a slightly modified version of "org.jboss.as.test.shared.ServerReload" as "org.wildfly.camel.test.common.ServerReload".
I hope this solution is lightweight enough.

jamesnetherton commented on itests/src/main/java/org/wildfly/camel/test/cxf/CxfWsCdiSecureExampleTest.java:
```diff
@@ -0,0 +1,183 @@
+                this.setManagementClient(managementClient);
+
+                ModelNode[] steps = new ModelNode[8];
+                steps[0] = ModelUtil.createOpNode(ADDRESS_SYSTEM_PROPERTY_TRUST_STORE, ModelDescriptionConstants.ADD);
```
We have a `DMRUtils` builder class which may be better than having to depend on external stuff.

jochenr replied:
Oh, sorry. I've overseen that;-(
But I just changed it:-)

jamesnetherton commented on itests/src/main/java/org/wildfly/camel/test/cxf/CxfWsCdiSecureExampleTest.java:
```diff
@@ -0,0 +1,183 @@
+                ModelNode result = executeOperation(compositeNode);
+
+
+                this.reload();
```
Do we definitely need to reload? If not, we can break the dependency on `ContainerResourceMgmtTestBase`.

jochenr replied:
Unfortunately yes, the reload is needed;-(

jamesnetherton commented on camel-cxf-jaxws-cdi-secure/pom.xml:
```diff
@@ -0,0 +1,125 @@
+	<artifactId>example-camel-cxf-jaxws-cdi-secure</artifactId>
+	<packaging>war</packaging>
+
+	<dependencies>
```
We should also add the following dependency as it avoids errors showing up in the Eclipse IDE:

```
        <dependency>
            <groupId>org.apache.taglibs</groupId>
            <artifactId>taglibs-standard-spec</artifactId>
            <scope>provided</scope>
        </dependency>
```

jochenr replied:
:white_check_mark:

jochenr commented:
Hi,
I just added #72 as reference for this request.

jochenr commented on camel-cxf-jaxws-cdi-secure/pom.xml:
```diff
@@ -0,0 +1,125 @@
+			<scope>provided</scope>
+		</dependency>
+
+		<!-- only for debug purpose -->
```
the org.picketbox dependency is needed  for "org.jboss.security.auth.callback.UsernamePasswordHandler" used in "org.wildfly.camel.examples.cxf.jaxws.JBossCallbackHandlerTlsCert"          

                <dependency>
			<groupId>org.picketbox</groupId>
			<artifactId>picketbox</artifactId>
			<scope>provided</scope>
		</dependency>

jochenr commented on camel-cxf-jaxws-cdi-secure/src/main/java/org/wildfly/camel/examples/cxf/jaxws/Application.java:
```diff
@@ -0,0 +1,165 @@
+
+    private static final String CXF_PRODUCER_ENDPOINT_ADDRESS = "https://localhost:8443/webservices/greeting-secure-cdi";
+
+    // actually the same as producer address, but here "https" doesn't work. Is this a bug?
```
yes this was fixed:
https://github.com/wildfly-extras/wildfly-camel/issues/1940
and works well.
Thank you

jochenr commented:
hopefully I implemented all your improvement requests. anything else I can do?
thanks,
Jochen

jamesnetherton commented:
Thanks. I'm pretty busy with other things at the moment, but I'll try to do a final review by the end of the week or early next week.

jochenr commented:
thank you very much, but there's no need to hurry.
I wanted to do this in August/September and had to time until now.....

jamesnetherton commented:
One final request, please can we extract the CLI commands from steps 3, 4 & 5 in the README to a CLI script. It'll make things simpler for folks who aren't familiar with the WildFly CLI. We'll also need a script to undo the changes.

jochenr commented:
I just added the CLI commands and added them to the POM. I tried to follow what's done in the jms example, but did two things diifferently:
1.) The "beforeDeployment" and "afterDeployment" of the deploy goal are deprcated, so I used the execute-commands goal as described in the plugin documentation

2.) I did something very ugly, since my script needs a ":reload" before the deployment: I added a sleep of 3 seconds, using the antrun-plugin.
Sorry for that, but di you have any other idea?
I found, this JIRA:
https://issues.jboss.org/browse/WFLY-208
so other people also found that ":reload(blocking=true)" is missing....

jamesnetherton commented:
ok to test

jamesnetherton commented:
Merged (with some small tweaks) - thanks for contributing :+1:
